src/extra/components/adq.py

```diff
@@ -0,0 +1,1318 @@
+        keyword argument.
+
+        The processing is parallelized via `pasha`.
+
+        Args:
+            data (ArrayLike): Input data to find edges on.
+            edge_func (Callable, optional): Edge finding method to run
+                on each train trace, extra.signal.dled by default.
+            max_edges (int, optional): Maximal number of edges per
+                train, 50 by default.
+            parallel (int or None, optional): Nunmber of parallel
+                processes to use, by default 10 or a quarter of all cores
+                whichever is lower. Any non-positive value or 1 disable
+                parallelization.
+
+            Any further keyword arguments are passed to the edge finding
+            method.
+
+        Returns:
+            result (pandas.DataFrame): Edge positions and pulse heights.
+        """
+
+        edges, amplitudes = self.find_edge_array(
+            data, False, False, edge_func, max_edges, parallel, **edge_kw)
+
+        if data.ndim > 2:
+            index = pd.MultiIndex.from_product(
+                [np.arange(x) for x in data.shape[:-1]],
+                names=[f'dim_{i}' for i in range(data.ndim - 1)])
+        else:
+            index = pd.Index(np.arange(data.shape[0]), name='dim_0')
+
+        return self._shape_edges(edges, amplitudes, index)
+
+    def find_edge_array(self, data, labelled=True, squeeze_edges=True,
+                        edge_func=None, max_edges=50, parallel=None,
+                        **edge_kw):
+        """Find signal edges as sparse array.
+
+        Alternative method to [find_edges][AdqRawChannel.find_edges]
+        returning the results as sparse array objects, using nan as
+        placeholder value.
+
+        Args:
+            data (ArrayLike): Input data to find edges on.
+            labelled (bool, optional): Whether data is returned as a
+                labelled xarray (default) or unlabelled ndarray.
+            squeeze_edges (bool, optional): Whether to minimize the edge
+                axis length to the maxinum number of edges found per
+                row, True by default.
+            edge_func (Callable, optional): Edge finding method to run
+                on each train trace, extra.signal.dled by default.
+            max_edges (int, optional): Maximal number of edges per
+                train, 1/5000 of trace length by default.
+            parallel (int or None, optional): Nunmber of parallel
+                processes to use, by default 10 or a quarter of all cores
+                whichever is lower. Any non-positive value or 1 disable
+                parallelization.
+
+            Any further keyword arguments are passed to the edge finding
+            method.
+
+        Returns:
+            edges, amplitudes (numpy.ndarray): Tuple of edge positions
+                and amplitudes, only if not labelled.
+
+            result (xarray.Dataset): Edge positions and pulse heights,
+                only
```

Comment:
```suggestion
                only if labelled.
```

And likewise for `train_edge_array` and `pulse_edge_array`.

Reply:
Oops, thanks.